addons/account/models/company.py

```diff
@@ -545,6 +560,144 @@ def _auto_balance_opening_move(self):
+            side = 'debit' if line.balance > 0.0 or line.amount_currency > 0.0 else 'credit'
+            account = line.account_id
+            corresponding_lines_per_account[(account, side)] |= line
+
+        line_ids = []
+        open_balance = 0.0
+        for (account, side), values in to_process.items():
+            balance = values['balance']
+            amount_currency = values['amount_currency']
+            currency = values['currency']
+            corresponding_lines = corresponding_lines_per_account[(account, side)]
+
+            if self.currency_id.is_zero(balance):
+                for line in corresponding_lines:
+                    open_balance -= line.balance
+                    line_ids.append(Command.delete(line.id))
+            elif corresponding_lines:
+                # Update the existing lines.
+                corresponding_line = corresponding_lines[0]
+                open_balance -= corresponding_line.balance
+                open_balance += balance
+                line_ids.append(Command.update(corresponding_line.id, {
+                    'balance': balance,
+                    'amount_currency': amount_currency,
+                    'currency_id': currency.id,
+                }))
+
+                # If more than one line on this account, remove the others.
+                for line in corresponding_lines[1:]:
+                    open_balance -= line.balance
+                    line_ids.append(Command.delete(line.id))
+            else:
+                # Create a new line.
+                open_balance += balance
+                line_ids.append(Command.create({
+                    'name': _("Opening balance"),
+                    'balance': balance,
+                    'amount_currency': amount_currency,
+                    'currency_id': currency.id,
+                    'account_id': account.id,
+                }))
+
+        # Auto-balance.
+        balancing_account = self.get_unaffected_earnings_account()
+        balancing_move_lines = opening_move.line_ids.filtered(lambda x: x.account_id == balancing_account)
+        for i, line in enumerate(balancing_move_lines):
+            open_balance -= line.balance
+            if i > 0:
+                line_ids.append(Command.delete(line.id))
+
+        balancing_move_line = balancing_move_lines[:1]
+        if balancing_move_line and self.currency_id.is_zero(open_balance):
+            line_ids.append(Command.delete(balancing_move_line.id))
+        elif balancing_move_lines:
+            line_ids.append(Command.update(balancing_move_line.id, {
+                'balance': -open_balance,
+                'amount_currency': -open_balance,
+            }))
+        else:
+            line_ids.append(Command.create({
+                'name': _("Automatic Balancing Line"),
+                'account_id': balancing_account.id,
+                'balance': -open_balance,
+                'amount_currency': -open_balance,
+            }))
```

Comment:
Doesn't this feel awfully identical to what is being done inside of the loop above?
Wouldn't a function make sense?

```python
def del_lines(lines):
    for line in lines:
        open_balance -= line.balance
        yield Command.delete(line.id)

def update_vals(account, side, balance):
    corresponding_lines = corresponding_lines_per_account[(account, side)]
    amount_currency = rate * balance
    open_balance += balance
    if self.currency_id.is_zero(balance):
        yield from del_lines(corresponding_lines)
    elif corresponding_lines:
        line_to_update = corresponding_line[0]
        open_balance -= line_to_update.balance
        yield Command.update(line_to_update.id, {
            'balance': balance,
            'amount_currency': amount_currency,
        })
        yield from del_lines(corresponding_line[1:])
    else:
        yield Command.create({
            'name':_("Automatic Balancing Line"),
            'account_id': account.id,
            'balance': balance,
            'amount_currency': amount_currency,
        })
            
open_balance = 0
commands = []
for account, (debit, credit) in to_update.items():
    if debit is not None:
        commands.extend(update_vals(account, 'debit', debit))
    if credit is not None:
        commands.extend(update_vals(account, 'credit', credit))

balancing_account = self.get_unaffected_earnings_account()
commands.extend(update_vals(balancing_account, 'debit', max(balance, 0)))
commands.extend(update_vals(balancing_account, 'credit', max(-balance, 0)))
```

Reply:
Just noticed that you might want to add a `label` param to `update_vals`

Reply:
@william-andre Your suggestion looks like an usine a gaz imo^^ sorry

Reply:
"Usine à gaz" is half your lines of code :sweat_smile:
I think you got the wrong definition of Usine à gaz :slightly_smiling_face: